includes/wc-admin-order-functions.php

```diff
@@ -16,40 +16,80 @@ function wc_admin_order_product_lookup_entry( $order_id ) {
 	global $wpdb;
 
 	$order = wc_get_order( $order_id );
-	if ( ! $order ) {
+	if ( ! $order || 'shop_order_refund' === $order->get_type() ) {
 		return;
 	}
 
-	foreach ( $order->get_items() as $order_item ) {
-		$wpdb->replace(
+	if ( 'refunded' === $order->get_status() ) {
+		$wpdb->delete(
 			$wpdb->prefix . 'wc_order_product_lookup',
-			array(
-				'order_item_id'         => $order_item->get_id(),
-				'order_id'              => $order->get_id(),
-				'product_id'            => $order_item->get_product_id( 'edit' ),
-				'variation_id'          => $order_item->get_variation_id( 'edit' ),
-				'customer_id'           => ( 0 < $order->get_customer_id( 'edit' ) ) ? $order->get_customer_id( 'edit' ) : null,
-				'product_qty'           => $order_item->get_quantity( 'edit' ),
-				'product_gross_revenue' => $order_item->get_subtotal( 'edit' ),
-				'date_created'          => date( 'Y-m-d H:i:s', $order->get_date_created( 'edit' )->getTimestamp() ),
-			),
-			array(
-				'%d',
-				'%d',
-				'%d',
-				'%d',
-				'%d',
-				'%d',
-				'%f',
-				'%s',
-			)
+			array( 'order_id' => $order->get_id() )
 		);
+		return;
+	}
+
+	$refunds = wc_admin_get_order_refund_items( $order );
+
+	foreach ( $order->get_items() as $order_item ) {
+		$quantity_refunded = isset( $refunds[ $order_item->get_id() ] ) ? $refunds[ $order_item->get_id() ] : 0;
+		if ( $quantity_refunded >= $order_item->get_quantity( 'edit' ) ) {
+			$wpdb->delete(
+				$wpdb->prefix . 'wc_order_product_lookup',
+				array( 'order_item_id' => $order_item->get_id() ),
+				array( '%d' )
+			);
+		} else {
+			$wpdb->replace(
+				$wpdb->prefix . 'wc_order_product_lookup',
+				array(
+					'order_item_id'         => $order_item->get_id(),
+					'order_id'              => $order->get_id(),
+					'product_id'            => $order_item->get_product_id( 'edit' ),
+					'variation_id'          => $order_item->get_variation_id( 'edit' ),
+					'customer_id'           => ( 0 < $order->get_customer_id( 'edit' ) ) ? $order->get_customer_id( 'edit' ) : null,
+					'product_qty'           => $order_item->get_quantity( 'edit' ) - $quantity_refunded,
+					'product_gross_revenue' => $order_item->get_subtotal( 'edit' ),
+					'date_created'          => date( 'Y-m-d H:i:s', $order->get_date_created( 'edit' )->getTimestamp() ),
+				),
+				array(
+					'%d',
+					'%d',
+					'%d',
+					'%d',
+					'%d',
+					'%d',
+					'%f',
+					'%s',
+				)
+			);
+		}
 	}
 }
 // TODO: maybe replace these with woocommerce_create_order, woocommerce_update_order, woocommerce_trash_order, woocommerce_delete_order, as clean_post_cache might be called in other circumstances and trigger too many updates?
 add_action( 'save_post', 'wc_admin_order_product_lookup_entry', 10, 1 );
+add_action( 'woocommerce_order_refunded', 'wc_admin_order_product_lookup_entry' );
```

Comment:
Side note, I know there is another issue that talks about using Action Scheduler to queue up jobs for lookup tables. #429 - not sure if we want to start that trend here or do it in a follow-up. Doesn't seem like there is too much heavy lifting going on here, but something worth considering.

Reply:
I tried to implement this but ran into an infinite loop with the scheduler.  Pinged @claudiosanches to see if he had any insight into the issue.

Reply:
My attempt is at https://github.com/woocommerce/wc-admin/pull/1050, if you want to have a look.

Reply:
I'm happy to get this merged into master if the below changes to the refund amount look okay if you want to handle the action scheduler changes in #1050.